// @fleekhq/fleek-cli/dist/services/authentication/mocks/login-response.mock.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateApiKeyErrorResponse = {
    data: {},
    errors: [
        {
            message: "Unexpected error"
        }
    ]
};
exports.generateApiKeyResponse = {
    data: {
        generateApiKey: {
            key: 'eAVISPni9Bhn92A3GGz6eQ==',
            secret: 'F7mVpSSaNnDfMakU2fVqErkTb9oVRJmxESm6IBJJ=='
        }
    }
};
exports.mockToken = {
    access_token: 'eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6InVZNzNlbXNVd3VLXzBqNnVTTzBIWUpiUUFoV3hKWlI0c1Y4ejhseVVjcVkifQ.eyJqdGkiOiJEYkF4cTFWVHhWRk5rU1pKYm5ZMWgiLCJzdWIiOiJvV1pkUW15a0pyeGtSTndCIiwiaWF0IjoxNTg3NDIyNTYyLCJleHAiOjE1ODc0MjQzNjIsInNjb3BlIjoib3BlbmlkIGVtYWlsIHByb2ZpbGUgdGVhbV9pbmZvIiwiaXNzIjoiaHR0cHM6Ly9kZXYtYXV0aC5mbGVlay5jbyIsImF1ZCI6InRlcm1pbmFsIn0.ntBCXNczDVaHPeYqep2yfmr9RKcpNLeqvTlLfvOaUV9Y9gpXY-IwPkmXmL75QT7Y-vvKJaN8hEQ1FA4QzKbviaC4PlUdmxskUtZY_4gi1hFgzJ-Jwq8CuYLHkGfDGwrQ4LpVBVop8PReH2Gggejci3IayUFyG1LhvJI-Ly5QuIwbN0ZI_mFWJsqSXVJwUsopijM9F0ysZhhSioUwz9IuHq96_MrQ6CkV9Y7WBdQfVAsoxk4rF_90g9KVvwCa2G1S0ZD7vO75qyXPGmb9PcRZdrsfELFE3hTAl0jbVN3zEb4G6zHY_Y4Lv70y_q39O7J5vbKiTo_LLEG2R211Z0ZbJQ',
    expires_in: '10000',
    id_token: 'eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6InVZNzNlbXNVd3VLXzBqNnVTTzBIWUpiUUFoV3hKWlI0c1Y4ejhseVVjcVkifQ.eyJqdGkiOiJEYkF4cTFWVHhWRk5rU1pKYm5ZMWgiLCJzdWIiOiJvV1pkUW15a0pyeGtSTndCIiwiaWF0IjoxNTg3NDIyNTYyLCJleHAiOjE1ODc0MjQzNjIsInNjb3BlIjoib3BlbmlkIGVtYWlsIHByb2ZpbGUgdGVhbV9pbmZvIiwiaXNzIjoiaHR0cHM6Ly9kZXYtYXV0aC5mbGVlay5jbyIsImF1ZCI6InRlcm1pbmFsIn0.ntBCXNczDVaHPeYqep2yfmr9RKcpNLeqvTlLfvOaUV9Y9gpXY-IwPkmXmL75QT7Y-vvKJaN8hEQ1FA4QzKbviaC4PlUdmxskUtZY_4gi1hFgzJ-Jwq8CuYLHkGfDGwrQ4LpVBVop8PReH2Gggejci3IayUFyG1LhvJI-Ly5QuIwbN0ZI_mFWJsqSXVJwUsopijM9F0ysZhhSioUwz9IuHq96_MrQ6CkV9Y7WBdQfVAsoxk4rF_90g9KVvwCa2G1S0ZD7vO75qyXPGmb9PcRZdrsfELFE3hTAl0jbVN3zEb4G6zHY_Y4Lv70y_q39O7J5vbKiTo_LLEG2R211Z0ZbJQ',
    scope: 'openid email profile team_info',
    session_state: '0VMVGNqVkN',
    state: '0VMVGNqVkN',
    token_type: 'Bearer',
};
